### PR TITLE
IC-1593 apply access restrictions for PPs accessing referrals

### PR DESCRIPTION
**Pull request by tomsmyers**
## What does this pull request do?

checks if the PP user has access to an SU before allowing them to view a referral.

please review commit by commit to have a look at the small changes to community api client code and referral controller.

## What is the intent behind these changes?

"block access for PP 'limited access offenders' (LAOs)"

### CODE REVIEW

**Review by mahalrmoj — APPROVED** [review]
LGTM

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/CommunityAPIService.kt [hunk 1]
```diff
@@ -26,6 +27,22 @@ interface CommunityAPIService {
...
+  private val communityAPIClient: CommunityAPIClient,
+) {
+  fun checkIfAuthenticatedDeliusUserHasAccessToOffender(authentication: JwtAuthenticationToken, crn: String): Boolean {
+    val userAccessPath = UriComponentsBuilder.fromPath(offenderAccessLocation)
+      .buildAndExpand(crn)
+      .toString()
+
+    val client = communityAPIClient.withCustomAuth(authentication)
+    val response = client.makeSyncGetRequest(userAccessPath, UserAccessResponse::class.java)
+    return !(response.userExcluded || response.userRestricted)
```
[the remark this answers is not in the document]
Comment (tomsmyers) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/CommunityAPIService.kt]: good point - but i think outside the scope of this?

[the remark this answers is not in the document]
Comment (tomsmyers) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/CommunityAPIService.kt]: log a warning?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/authorization/ReferralAccessChecker.kt [hunk 2]
```diff
@@ -39,7 +42,10 @@ class ReferralAccessChecker(
     }
   }
 
-  private fun forProbationPractitionerUser(referral: Referral, user: AuthUser) {
-    // currently no access restrictions on PP users
+  private fun forProbationPractitionerUser(referral: Referral, user: AuthUser, authentication: JwtAuthenticationToken) {
+    val hasAccess = communityApiOffenderService.checkIfAuthenticatedDeliusUserHasAccessToOffender(authentication, referral.serviceUserCRN)
+    if (!hasAccess) {
+      throw AccessError(errorMessage, listOf("probation practitioner is excluded from access to this service user"))
```
[the remark this answers is not in the document]
Comment (tomsmyers) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/authorization/ReferralAccessChecker.kt]: i have updated this message in a later commit i think 
 
 it reads "user is not allowed to access records related to this service user"

**Pull request merged**